File: client/lib/api/v2/hooks/assistant.ts
/**
 * V2 API hooks for assistant
 * Centralized data fetching for assistant chats
 */

import { api } from "@/lib/api/fetcher";
import { assistantChatsFullKeys } from "@/lib/api/v2/keys";
import { useQuery } from "@tanstack/react-query";

export interface AssistantChatFullResponse {
  chat: {
    id: string;
    createdAt: string;
    updatedAt: string;
    profileId: string;
    title: string;
    traceId: string | null;
  } | null;
  messages: Array<{
    id: string;
    createdAt: string;
    updatedAt: string;
    completedAt: string | null;
    chatId: string;
    role: "user" | "assistant";
    content: string;
    completed: boolean;
  }>;
  toolCalls: Array<{
    id: string;
    createdAt: string;
    updatedAt: string;
    completedAt: string | null;
    chatId: string;
    toolName: string;
    toolType: "create" | "read" | "update" | "delete";
    toolArguments: unknown;
    toolResult: unknown;
    completed: boolean;
  }>;
  allChats: Array<{
    id: string;
    createdAt: string;
    updatedAt: string;
    profileId: string;
    title: string;
    traceId: string | null;
  }>;
}

/**
 * Hook to fetch complete assistant chat data with all related entities
 * @param chatId - Optional chat ID. If undefined, only fetches the chats list
 * @param profileId - Profile ID to fetch all chats for dropdown
 */
export function useAssistantChatFull(
  chatId: string | undefined,
  profileId: string,
  enabled = true
) {
  return useQuery<AssistantChatFullResponse>({
    queryKey: assistantChatsFullKeys.detail(chatId, profileId),
    queryFn: async () => {
      if (!chatId) {
        // Fetch only the chats list for new chat state
        return api<AssistantChatFullResponse>(
          `/api/v2/assistant/chats/list/${profileId}`
        );
      }
      // Fetch full chat data
      return api<AssistantChatFullResponse>(
        `/api/v2/assistant/chats/${chatId}/full?profile_id=${profileId}`
      );
    },
    enabled: enabled && !!profileId && profileId !== "",
    staleTime: 1000, // 1 second - allow WebSocket updates to trigger refetch
    refetchOnWindowFocus: false, // Don't refetch on window focus since we have WebSocket updates
  });
}

File: server/app/api/v2/assistant.py
"""Assistant v2 API endpoints."""

from typing import Annotated, Any, Dict
from uuid import UUID

import asyncpg  # type: ignore
from app.db import get_db
from app.queries.assistant_queries import get_assistant_chat_full_data
from fastapi import APIRouter, Depends, HTTPException, Query

router = APIRouter(prefix="/assistant", tags=["assistant"])


@router.get("/chats/{chat_id}/full")
async def get_assistant_chat_full(
    chat_id: UUID,
    profile_id: Annotated[UUID, Query()],
    conn: Annotated[asyncpg.Connection, Depends(get_db)],
) -> Dict[str, Any]:
    """Get complete assistant chat data with all related entities."""
    try:
        return await get_assistant_chat_full_data(conn, chat_id, profile_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/chats/list/{profile_id}")
async def get_assistant_chats_list(
    profile_id: UUID,
    conn: Annotated[asyncpg.Connection, Depends(get_db)],
) -> Dict[str, Any]:
    """Get all chats for a profile (for new chat state without chat_id)."""
    try:
        # Pass None as chat_id to only fetch the chats list
        return await get_assistant_chat_full_data(conn, None, profile_id)  # type: ignore
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

